File: 0145-binary-tree-postorder-traversal/iterative-depth-first-search/python/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        stack = [root]

        while stack:
           cur = stack.pop()
           if cur:
               result.append(cur.val)
               stack.append(cur.left)    # Push left first so that right sits on top and gets popped/processed first
               stack.append(cur.right)
        return result[::-1]          # Reverse to convert root-right-left into left-right-root (i.e. postorder)

File: 0145-binary-tree-postorder-traversal/iterative-depth-first-search/java/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;

        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        
        while (!stack.isEmpty()) {
            TreeNode cur = stack.pop();
            if (cur != null) {
                result.add(cur.val); 
                if (cur.left != null)  stack.push(cur.left);    // Push left first so that right sits on top and gets popped/processed first
                if (cur.right != null) stack.push(cur.right);
            }
        }
        
        Collections.reverse(result); // Reverse to convert root-right-left into left-right-root (i.e. postorder)
        return result;
    }
}
